### PR TITLE
Indicate if profile user follows logged in user

### PR DESCRIPTION
**Pull request by dzamie**
(Suggested in Issue #8)
If a viewed user is following the current user, "Following You" will
appear to the left of "Follow" on their user page.
Added code to support an icon for "Following You", which could be placed
starting at (0, 260) of "./images/icons-20.png" (Current icons-20.png is
60x260, so an extra row would have to be added).

### CODE REVIEW

**Comment by weykent** [comment]
Nice! Thank you!

File: templates/common/user_tools.html [hunk 1]
```diff
@@ -44,6 +44,10 @@ <h2 class="username">
       <ul id="user-actions" class="toolset clear">
 
         $if not relationship['is_self']:
+		  <li>
+		    $if relationship['followby']:
+			  <span class="icon icon-20 icon-followed"></span>Following You
+		  </li>
```
Comment (weykent) [inline templates/common/user_tools.html]: This code doesn't follow the indentation in the rest of the file. I can't check right this moment, but I don't think the templates use tabs at all.

File: weasyl/profile.py [hunk 2]
```diff
@@ -290,14 +290,16 @@ def select_relation(userid, otherid):
...
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%i, %i) AND settings ~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (otherid, userid) = (%i, %i)))
+    """, [userid, otherid, userid, otherid, userid, otherid, userid, otherid, userid, otherid, otherid, userid], ["single"])
 
     return {
         "follow": query[0],
         "friend": query[1],
         "ignore": query[2],
         "friendreq": query[3],
+        "followby": query[4],
```
Comment (strain-113) [inline weasyl/profile.py]: wording nit: I think "follower" may be clearer terminology here.

File: static/public.css [hunk 3]
```diff
@@ -1169,6 +1169,7 @@ a:hover .icon-arrowDown, button:hover .icon-arrowDown { background-position: -20
 .icon-report { background-position: 0 -240px; }
 a:hover .icon-report, button:hover .icon-report { background-position: -20px -240px; }
 .active .icon-report { background-position: -40px -240px; }
+.icon-followed { background-position: 0 -260px; }
```
Comment (strain-113) [inline static/public.css]: wording nit: I think '.icon-follower' is clearer terminology.

File: weasyl/profile.py [hunk 4]
```diff
@@ -290,14 +290,16 @@ def select_relation(userid, otherid):
             (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = (%i, %i))),
             (SELECT EXISTS (SELECT 0 FROM frienduser WHERE userid IN (%i, %i) AND otherid IN (%i, %i) AND settings !~ 'p')),
             (SELECT EXISTS (SELECT 0 FROM ignoreuser WHERE (userid, otherid) = (%i, %i))),
-            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%i, %i) AND settings ~ 'p'))
-    """, [userid, otherid, userid, otherid, userid, otherid, userid, otherid, userid, otherid], ["single"])
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%i, %i) AND settings ~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (otherid, userid) = (%i, %i)))
+    """, [userid, otherid, userid, otherid, userid, otherid, userid, otherid, userid, otherid, otherid, userid], ["single"])
```
Comment (strain-113) [inline weasyl/profile.py]: If this is replaced with named parameters (and the query statement switched to d.engine.execute), it will be easier to read and maintain than the long list of getting userid and otherid in the right order.

File: templates/common/user_tools.html (hunk 1 above)
Comment (dzamie) [inline templates/common/user_tools.html]: Possibly not. The editor I used (and for some reason, GitHub's preview) showed them as 2-space indents, so I never noticed anything off and assumed for the most part that everything was tabbed.
I'm kind of new to using github; is there a way I can change this, or do I have to close the request and make a new one?

File: static/public.css (hunk 3 above)
Comment (dzamie) [inline static/public.css]: I agree. I had somehow completely forgotten the word "follower" exists.

File: templates/common/user_tools.html (hunk 1 above)
Comment (strain-113) [inline templates/common/user_tools.html]: You can make another commit on top of this one and push it to the branch. Just 'fix whitespace' or whatever. PRs will track changes to the branch.

You can also do a git commit --amend and force push to the branch, but GitHub's support for that flow is less stable.

File: weasyl/profile.py (hunk 4 above)
Comment (dzamie) [inline weasyl/profile.py]: I think I've managed to do something like that. It still uses d.execute (unfamiliar with engine), but query is formatted first, and it looks much nicer.

File: weasyl/profile.py [hunk 5]
```diff
@@ -285,19 +285,22 @@ def select_relation(userid, otherid):
...
-            (SELECT EXISTS (SELECT 0 FROM ignoreuser WHERE (userid, otherid) = (%i, %i))),
-            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%i, %i) AND settings ~ 'p'))
-    """, [userid, otherid, userid, otherid, userid, otherid, userid, otherid, userid, otherid], ["single"])
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = ({user_id}, {other_id}))),
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE userid IN ({user_id}, {other_id}) AND otherid IN ({user_id}, {other_id}) AND settings !~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM ignoreuser WHERE (userid, otherid) = ({user_id}, {other_id}))),
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = ({user_id}, {other_id}) AND settings ~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = ({other_id}, {user_id})))
+    """.format(user_id=userid, other_id=otherid)
+    query = d.execute(query, None, ["single"])
```
Comment (charmander) [inline weasyl/profile.py]: `d.engine.execute` allows for proper parameters:

```python
result = d.engine.execute("""
    SELECT
        (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = (%(user)s, %(other)s))),
        (SELECT EXISTS (SELECT 0 FROM frienduser WHERE userid IN (%(user)s, %(other)s) AND otherid IN (%(user)s, %(other)s) AND settings !~ 'p')),
        (SELECT EXISTS (SELECT 0 FROM ignoreuser WHERE (userid, otherid) = (%(user)s, %(other)s))),
        (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%(user)s, %(other)s) AND settings ~ 'p')),
        (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = (%(user)s, %(other)s)))
""", user=userid, other=otherid).first()
```

With this, you can also give each field an alias (`(SELECT EXISTS …) AS follow`) and use it (`"follow": result.follow`).

Comment (dzamie) [inline weasyl/profile.py]: It still works with the following code, successfully getting `"ignore"` status from `result[2]`, right?

File: weasyl/profile.py [hunk 6]
```diff
@@ -285,19 +285,21 @@ def select_relation(userid, otherid):
...
-    """, [userid, otherid, userid, otherid, userid, otherid, userid, otherid, userid, otherid], ["single"])
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = (%(user)s, %(other)s))),
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE userid IN (%(user)s, %(other)s) AND otherid IN (%(user)s, %(other)s) AND settings !~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM ignoreuser WHERE (userid, otherid) = (%(user)s, %(other)s))),
+            (SELECT EXISTS (SELECT 0 FROM frienduser WHERE (userid, otherid) = (%(user)s, %(other)s) AND settings ~ 'p')),
+            (SELECT EXISTS (SELECT 0 FROM watchuser WHERE (userid, otherid) = (%(other)s, %(user)s)))
+    """, user=userid, other=otherid).first()
 
     return {
         "follow": query[0],
```
Comment (charmander) [inline weasyl/profile.py]: In the query, could you add `AS follow`, `AS friend`, `AS ignore`, `AS friendreq`, `AS follower` after each of the subqueries, please? Then this can become:

```python
return dict(
    query,
    is_self=False)
```

**Comment by strain-113** [comment]
This is looking good. Can you add a new unit test to test the output of the new `select_profile()` for various relationships? (e.g. she follows me, I follow her, we follow each other, we don't follow each other at all, etc.)

After that I think it's just about UI?

**Comment by dzamie** [comment]
@strain-113 I don't see what it would test. The new `follower` stuff involves looking at two users (`userid` and `otherid`), whereas `select_profile()` seems to only look at one user. The only functionality added was to see if `otherid` follows `userid` rather than only the other way. Nothing I saw in `test_followuser.py` would result in any test other than swapping around variable names.

**Comment by strain-113** [comment]
Apologies: That was a typo. I meant to write `select_relation()`. That is, write tests that given various calls to `followuser` test the output of `select_relation(userid, otherid)`.

**Comment by dzamie** [comment]
Odd, it stops working when I try to add the tests, but I can't figure out what the older version of Python doesn't implement that breaks it; the current version runs fine.

**Comment by weykent** [comment]
Only libweasyl, not weasyl, supports python 3.5, so your tests are ignored under 3.5. Your tests are failing because you haven't imported the function you're calling. Running the linter should catch this.

**Comment by strain-113** [comment]
Nit: commit messages should adhere to https://github.com/Weasyl/weasyl/wiki/Style-guide-and-Best-Practices#commit-messages

No need to rewrite existing ones.

File: weasyl/test/test_followuser.py [hunk 7]
```diff
@@ -80,5 +92,20 @@ def test_remove(self, followuser_remove):
...
+        user2 = db_utils.create_user()
+
+        # user1 watches user2
+        followuser.insert(user1, user2)
+        # user2 watches user1
+        followuser.insert(user2, user1)
+        self.assertTrue(profile.select_relation(user1, user2)["follow"])
+        self.assertTrue(profile.select_relation(user2, user1)["follower"])
+        self.assertTrue(profile.select_relation(user2, user1)["follow"])
+        self.assertTrue(profile.select_relation(user1, user2)["follower"])
```
Comment (charmander) [inline weasyl/test/test_followuser.py]: A trailing newline is missing here.

Comment (dzamie) [inline weasyl/test/test_followuser.py]: Like so? Or was there one to put inside the test that I missed?

Comment (charmander) [inline weasyl/test/test_followuser.py]: Nope, that’s right!

**Comment by charmander** [comment]
Is this how it’s supposed to look?

screenshot; the indicator is small and misaligned, with the wrong icon

**Comment by charmander** [comment]
Is this how it’s supposed to look?

![screenshot; the indicator is misaligned][1]

[1]: https://cloud.githubusercontent.com/assets/1889843/16179919/bc56a492-3629-11e6-9ee6-a036520e1fd6.png

**Comment by charmander** [comment]
It also breaks the layout at widths of 20–22 and 27–32 ems.

**Comment by dzamie** [comment]
Huh, the text formatting looks identical to Follow/Friend/Message/Ignore for me. I'm trying to put in a workaround.
Not sure how to fix "This branch has conflicts" message on here; syncing, editing, and pushing to dzamie/master didn't seem to do anything.

**Comment by dzamie** [comment]
I think I found the error; the CSS only works in <li><a> tags, so I'm trying to put the addition in href-less <a> tags as a workaround. For some reason, I can't resolve the conflicts GitHub is telling me about.

File: templates/common/user_tools.html [hunk 8]
```diff
@@ -46,7 +46,8 @@ <h2 class="username">
         $if not relationship['is_self']:
           <li>
             $if relationship['follower']:
-              <span class="icon icon-20 icon-follower"></span>Following You
+              <!-- The <a> tag triggers the CSS to match the font. -->
+              <a><span class="icon icon-20 icon-follower"></span>Following You</a>
```
Comment (charmander) [inline templates/common/user_tools.html]: No. Modify the CSS, please.

**Comment by charmander** [comment]
Don’t worry about fixing conflicts; there aren’t really any.

**Comment by charmander** [comment]
Okay, I’m going to merge the Python changes here and we can look at the CSS and template changes separately. Thanks for doing this!

**Comment by charmander** [comment]
Merged in 07e1744ac46ec4a8c8aba29de45462677c2087d2.
